harness/determined/estimator/_estimator_context.py

```diff
@@ -133,3 +133,98 @@ class EstimatorExperimentalContext(_data_layer.DataLayerContext):
+
+        @functools.wraps(f)
+        def decorated(*arg: Any, **kwarg: Any) -> tf.Operation:
+            with tf.compat.v1.control_dependencies(self._allgather_ops):
+                new_op = f(*arg, **kwarg)
+            self._allgather_ops.append(new_op)
+            return new_op
+
+        return decorated
+
+    def make_metric(
+        self,
+        metric: Any,
+        reducer: Union[Callable[[List[Any]], Any], "estimator.MetricReducer"],
+        name: str = "custom_metric",
+    ) -> tf.keras.metrics.Metric:
+        """
+        Return an estimator-compatible metric which will be calculated properly, even during
+        distributed training.
+
+        Unfortunately, during distributed evaluation, many types of metrics calculated via
+        ``tf.metrics`` or ``tf.keras.metrics`` cannot be aggregated properly from the per-slot
+        final metrics calculated by each separate Estimator replica.  One example is
+        ``tf.metrics.auc``, where the ROC AUC calculated over predictions and labels from a full
+        dataset cannot be derived from a set of ROC AUC metrics evaluated over the shards of a
+        dataset.  However, with ``make_metric``, the ROC AUC could be calculated in distributed
+        training by calling ``sklearn.metrics.roc_auc_score`` in a custom ``reducer`` function.
+
+        Determined solves this problem by offering customizable metrics which are
+        Estimator-compatible.
+
+        The ``metric`` input can be a tensor, a list of tensors, or a dictionary of tensors.
+
+        The ``reducer`` should be either a single function that can calculate the metric from a
+        list of the per-batch values of ``metric``, or it can be an instance of a
+        :class:`det.estimator.MetricReducer<determined.estimator.MetricReducer>`.
+
+        Example usage where ``reducer`` is a function:
+
+        .. code-block:: python
+
+           def my_mean_reducer(all_batch_metrics):
+               # Use hstack in case not all batches are equal length.
+               return np.mean(np.hstack(all_batch_metrics))
+
+           def my_estimator_model_function(features, labels, mode):
+               ...
+               if mode == tf.estimator.ModeKeys.EVAL:
+
+                   my_avg_prediction = context.experimental.make_metric(
+                        metric=predictions, reducer=my_mean_reducer
+                   )
+
+                   return tf.estimator.EstimatorSpec(
+                       mode,
+                       loss=loss,
+                       eval_metric_ops={"my_avg_prediction": my_avg_prediction},
+                   )
+
+        The primary motivation for passing a function as the reducer is simplicity.  Metrics from
+        all batches will be buffered in memory and passed over the network where they will be
+        reduced all at once.  This is probably fine for scalar metrics, or on validation datasets
```

Comment:
Rephrase: `This is probably fine`

Reply:
rewrote the whole second half of this paragraph, let me know if you like it.

Reply:
Yeah i like the current version, left one comment relate to the last paragraph.